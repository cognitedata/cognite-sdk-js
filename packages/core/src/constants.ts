// Copyright 2020 Cognite AS

/** @hidden */
export const API_VERSION: string = 'v1';
/** @hidden */
export const DEFAULT_CLUSTER = 'api';
/** @hidden */
export const DEFAULT_DOMAIN = 'cognitedata.com';
/** @hidden */
export const BASE_URL: string = `https://${DEFAULT_CLUSTER}.${DEFAULT_DOMAIN}`;

/** @hidden */
export const AUTHORIZATION_HEADER = 'Authorization';
/** @hidden */
export const API_KEY_HEADER = 'api-key';
/** @hidden */
export const X_CDF_APP_HEADER = 'x-cdp-app';
/** @hidden */
export const X_CDF_SDK_HEADER = 'x-cdp-sdk';
/** @hidden */
export const X_REQUEST_ID = 'X-Request-Id';

/** @hidden */
export const LOCAL_STORAGE_PREFIX = '@cognite/sdk:';
