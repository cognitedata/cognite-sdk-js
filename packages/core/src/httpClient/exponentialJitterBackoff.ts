  /**
   * Calculates the delay in milliseconds for the next retry attempt. Uses "exponential backoff with 
   * full jitter" algorithm.
   * 
   * @see https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
   * 
   * @param retryCount - The number of retry attempts made so far.
   * @returns The delay in milliseconds for the next retry attempt.
   */
export class ExponentialJitterBackoff {
  private readonly baseDelayMs: number;
  private readonly maxDelayMs: number;
  private readonly randomCallback: typeof Math.random;

  /**
   * Initializes a new instance of the ExponentialJitterBackoff class.
   * @param baseDelayMs - The base delay in milliseconds. This will be the average delay for the first retry attempt.
   * @param maxDelayMs - The maximum delay in milliseconds. This will be the maximum delay for the last retry attempt.
   * @param randomCallback - The callback to use for generating a random number between 0 and 1. Used by test only.
   */
  constructor(
    baseDelayMs: number = 100,
    maxDelayMs: number = 5000,
    randomCallback: typeof Math.random = Math.random,
  ) {
    this.baseDelayMs = baseDelayMs;
    this.maxDelayMs = maxDelayMs;
    this.randomCallback = randomCallback;
  }

  public calculateDelayInMs(retryCount: number) {
    const exponentialDelay = this.baseDelayMs * (2 ** (retryCount + 1));
    const cappedDelay = Math.min(exponentialDelay, this.maxDelayMs);

    return Math.floor(this.randomCallback() * cappedDelay);
  }
}
