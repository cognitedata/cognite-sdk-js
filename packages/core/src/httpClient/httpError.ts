// Copyright 2020 Cognite AS
import type { HttpHeaders } from './httpHeaders';

export interface HttpErrorData {
  error: {
    code: number;
    message: string;
    isAutoRetryable?: boolean;
    missing?: object[];
    duplicated?: object[];
    forbidden?: object[];
    extra?: unknown;
  };
}

export class HttpError extends Error {
  /** @hidden */
  constructor(
    public status: number,
    public data: HttpErrorData,
    public headers: HttpHeaders
  ) {
    super(`Request failed | status code: ${status}`);
    Object.setPrototypeOf(this, HttpError.prototype); // https://stackoverflow.com/questions/51229574/why-instanceof-returns-false-for-a-child-object-in-javascript
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(this.message).stack;
    }
  }
}
