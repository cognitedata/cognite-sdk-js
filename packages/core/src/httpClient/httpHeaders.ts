export interface HttpHeaders {
  [key: string]: string;
}
