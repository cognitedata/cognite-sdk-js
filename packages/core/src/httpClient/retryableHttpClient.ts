// Copyright 2020 Cognite AS

import isFunction from 'lodash/isFunction';
import { sleepPromise } from '../utils';
import {
  BasicHttpClient,
  type HttpRequest,
  type HttpRequestOptions,
  type HttpResponse,
} from './basicHttpClient';
import { MAX_RETRY_ATTEMPTS, type RetryValidator } from './retryValidator';

export class RetryableHttpClient extends BasicHttpClient {
  private static calculateRetryDelayInMs(retryCount: number) {
    const INITIAL_RETRY_DELAY_IN_MS = 250;
    return INITIAL_RETRY_DELAY_IN_MS + ((2 ** retryCount - 1) / 2) * 1000;
  }

  constructor(
    baseUrl: string,
    private retryValidator: RetryValidator
  ) {
    super(baseUrl);
  }

  public get<ResponseType>(
    path: string,
    options: RetryableHttpRequestOptions = {}
  ) {
    return super.get<ResponseType>(path, options);
  }

  public post<ResponseType>(
    path: string,
    options: RetryableHttpRequestOptions = {}
  ) {
    return super.post<ResponseType>(path, options);
  }

  public put<ResponseType>(
    path: string,
    options: RetryableHttpRequestOptions = {}
  ) {
    return super.put<ResponseType>(path, options);
  }

  public delete<ResponseType>(
    path: string,
    options: RetryableHttpRequestOptions = {}
  ) {
    return super.delete<ResponseType>(path, options);
  }

  public patch<ResponseType>(
    path: string,
    options: RetryableHttpRequestOptions = {}
  ) {
    return super.patch<ResponseType>(path, options);
  }

  protected async preRequest(
    request: RetryableHttpRequest
  ): Promise<RetryableHttpRequest> {
    return super.preRequest(request);
  }

  protected async postRequest<T>(
    response: HttpResponse<T>,
    request: RetryableHttpRequest,
    mutatedRequest: RetryableHttpRequest
  ): Promise<HttpResponse<T>> {
    return super.postRequest<T>(response, request, mutatedRequest);
  }

  protected async rawRequest<ResponseType>(
    request: RetryableHttpRequest
  ): Promise<HttpResponse<ResponseType>> {
    let retryCount = 0;
    // eslint-disable-next-line no-constant-condition
    while (true) {
      const response = await super.rawRequest<ResponseType>(request);

      const retryValidator = isFunction(request.retryValidator)
        ? request.retryValidator
        : this.retryValidator;

      const shouldRetry =
        retryCount < MAX_RETRY_ATTEMPTS &&
        request.retryValidator !== false &&
        retryValidator(request, response, retryCount);
      if (!shouldRetry) {
        return response;
      }
      const delayInMs = RetryableHttpClient.calculateRetryDelayInMs(retryCount);
      await sleepPromise(delayInMs);
      retryCount++;
    }
  }

  protected async request<ResponseType>(request: RetryableHttpRequest) {
    return super.request<ResponseType>(request);
  }
}

export type RetryableHttpRequest = HttpRequest &
  HttpRequestRetryValidatorOptions;

export type RetryableHttpRequestOptions = HttpRequestOptions &
  HttpRequestRetryValidatorOptions;

type HttpRequestRetryValidatorOptions = {
  retryValidator?: false | RetryValidator;
};
