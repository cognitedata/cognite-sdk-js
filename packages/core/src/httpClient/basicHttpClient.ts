// Copyright 2020 Cognite AS

import fetch, { Response } from 'cross-fetch';
import { stringify } from 'query-string';
import { isJson } from '../utils';
import { HttpError } from './httpError';
import { DEFAULT_DOMAIN } from '../constants';

export class BasicHttpClient {
  private static validateStatusCode(status: number) {
    return status >= 200 && status < 300;
  }
  private static throwCustomErrorResponse<T>(response: HttpResponse<T>) {
    throw new HttpError(response.status, response.data, response.headers);
  }

  private static jsonResponseHandler<ResponseType>(
    res: Response
  ): Promise<ResponseType> {
    return res.json() as Promise<ResponseType>;
  }

  private static textResponseHandler<ResponseType>(
    res: Response
  ): Promise<ResponseType> {
    return (res.text() as unknown) as Promise<ResponseType>;
  }

  private static arrayBufferResponseHandler<ResponseType>(
    res: Response
  ): Promise<ResponseType> {
    return (res
      .blob()
      .then(blob => new Response(blob).arrayBuffer()) as unknown) as Promise<
      ResponseType
    >;
  }

  private static getResponseHandler<ResponseType>(
    responseType?: HttpResponseType
  ): ResponseHandler<ResponseType> {
    switch (responseType) {
      case HttpResponseType.ArrayBuffer:
        return BasicHttpClient.arrayBufferResponseHandler;
      case HttpResponseType.Text:
        return BasicHttpClient.textResponseHandler;
      default:
        return BasicHttpClient.jsonResponseHandler;
    }
  }

  private static convertFetchHeaders(fetchHeaders: Headers) {
    const headers: HttpHeaders = {};
    fetchHeaders.forEach((value, key) => {
      headers[key] = value;
    });
    return headers;
  }

  private static transformRequestBody(data: any) {
    const isJSONStringifyable = isJson(data);
    if (isJSONStringifyable) {
      return JSON.stringify(data, null, 2);
    }
    return data;
  }

  private static resolveUrl(baseUrl: string, path: string) {
    const trimmedBaseUrl = baseUrl.replace(/\/$/, '');
    const pathWithPrefix = (path[0] === '/' ? '' : '/') + path;
    return trimmedBaseUrl + pathWithPrefix;
  }

  private defaultHeaders: HttpHeaders = {};

  /**
   * A basic http client with the option of adding default headers,
   * and the option of using a baseUrl.
   * Note: If the path given starts with `https://`, the baseUrl is ignored.
   *
   * It handles query parameters, and request body data which gets encoded as json if
   * needed.
   *
   * See HttpRequest.
   *
   * @param baseUrl
   */
  constructor(protected baseUrl: string) {}

  public setDefaultHeader(name: string, value: string) {
    this.defaultHeaders[name] = value;
    return this;
  }

  public getDefaultHeaders(): HttpHeaders {
    return { ...this.defaultHeaders };
  }

  public setBaseUrl = (baseUrl: string) => {
    this.baseUrl = baseUrl;
  };

  public getBaseUrl() {
    return this.baseUrl;
  }

  public setCluster = (cluster: string) => {
    this.baseUrl = `https://${cluster}.${DEFAULT_DOMAIN}`;
  };

  public get<ResponseType>(path: string, options: HttpRequestOptions = {}) {
    return this.request<ResponseType>({
      ...options,
      path,
      method: HttpMethod.Get,
    });
  }

  // TODO: responseType -> factory pattern
  public post<ResponseType>(path: string, options: HttpRequestOptions = {}) {
    return this.request<ResponseType>({
      ...options,
      path,
      method: HttpMethod.Post,
    });
  }

  public put<ResponseType>(path: string, options: HttpRequestOptions = {}) {
    return this.request<ResponseType>({
      ...options,
      path,
      method: HttpMethod.Put,
    });
  }

  public delete<ResponseType>(path: string, options: HttpRequestOptions = {}) {
    return this.request<ResponseType>({
      ...options,
      path,
      method: HttpMethod.Delete,
    });
  }

  public patch<ResponseType>(path: string, options: HttpRequestOptions = {}) {
    return this.request<ResponseType>({
      ...options,
      path,
      method: HttpMethod.Patch,
    });
  }

  protected async preRequest(request: HttpRequest): Promise<HttpRequest> {
    const populatedHeaders = this.populateDefaultHeaders(request.headers);
    return {
      ...request,
      headers: populatedHeaders,
    };
  }

  protected async postRequest<T>(
    response: HttpResponse<T>,
    _: HttpRequest // eslint-disable-line
  ): Promise<HttpResponse<T>> {
    const requestIsOk = BasicHttpClient.validateStatusCode(response.status);
    if (!requestIsOk) {
      BasicHttpClient.throwCustomErrorResponse(response);
    }
    return response;
  }

  protected populateDefaultHeaders(headers: HttpHeaders = {}) {
    return {
      ...this.defaultHeaders,
      ...headers,
    };
  }

  protected async rawRequest<ResponseType>(
    request: HttpRequest
  ): Promise<HttpResponse<ResponseType>> {
    const url = this.constructUrl(request.path, request.params);
    const headers: HttpHeaders = {
      Accept: 'application/json',
      ...request.headers,
    };
    let body = request.data;
    if (isJson(body)) {
      body = BasicHttpClient.transformRequestBody(body);
      headers['Content-Type'] = 'application/json';
    }
    const res = await fetch(url, {
      body,
      method: request.method,
      headers,
    });
    const responseHandler = BasicHttpClient.getResponseHandler<ResponseType>(
      request.responseType
    );
    const data = await responseHandler(res);
    return {
      headers: BasicHttpClient.convertFetchHeaders(res.headers),
      status: res.status,
      data,
    };
  }

  protected async request<ResponseType>(request: HttpRequest) {
    const mutatedRequest = await this.preRequest(request);
    const rawResponse = await this.rawRequest<ResponseType>(mutatedRequest);
    return this.postRequest(rawResponse, request);
  }

  private constructUrl(path: string, params: HttpQueryParams = {}) {
    let url = path;
    const hasQueryParams = Object.keys(params).length > 0;
    if (hasQueryParams) {
      url += `?${stringify(params)}`;
    }
    const urlContainsHost = url.match(/^https?:\/\//i) !== null;
    if (urlContainsHost) {
      return url;
    }
    return BasicHttpClient.resolveUrl(this.baseUrl, url);
  }
}

export interface HttpRequest extends HttpRequestOptions {
  path: string;
  method: HttpMethod;
}

export interface HttpRequestOptions {
  data?: any;
  params?: HttpQueryParams;
  headers?: HttpHeaders;
  responseType?: HttpResponseType;
  /**
   * Set this to 'true' if you want to send credentials (api-key, access token) with the request to other domains than the specified base url.
   */
  withCredentials?: boolean;
}

export interface HttpResponse<T> {
  data: T;
  status: number;
  headers: HttpHeaders;
}

export enum HttpMethod {
  Get = 'GET',
  Post = 'POST',
  Put = 'PUT',
  Delete = 'DELETE',
  Patch = 'PATCH',
}

export type HttpResponseType = 'json' | 'arraybuffer' | 'text';

export const HttpResponseType = {
  Json: 'json' as HttpResponseType,
  ArrayBuffer: 'arraybuffer' as HttpResponseType,
  Text: 'text' as HttpResponseType,
};

export interface HttpQueryParams {
  [key: string]: any;
}

export interface HttpHeaders {
  [key: string]: string;
}

type ResponseHandler<ResponseType> = (res: Response) => Promise<ResponseType>;

export type HttpCall = <ResponseType>(
  path: string,
  options?: HttpRequestOptions
) => Promise<HttpResponse<ResponseType>>;
