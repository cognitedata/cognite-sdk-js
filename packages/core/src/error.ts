// Copyright 2020 Cognite AS
import { X_REQUEST_ID } from './constants';
import type { HttpError } from './httpClient/httpError';
import { getHeaderField } from './httpClient/httpHeaders';

export class CogniteError extends Error {
  public status: number;
  public errorMessage: string;
  public requestId?: string;
  public missing?: object[];
  public duplicated?: unknown[];
  public extra?: unknown;
  /** @hidden */
  constructor(
    errorMessage: string,
    status: number,
    requestId?: string,
    otherFields: {
      missing?: object[];
      duplicated?: unknown[];
      extra?: unknown;
    } = {},
  ) {
    let message = `${errorMessage} | code: ${status}`;
    if (requestId) {
      message += ` | X-Request-ID: ${requestId}`;
    }
    const { missing, duplicated, extra } = otherFields;
    if (missing || duplicated) {
      message += `\n${JSON.stringify(otherFields, null, 2)}`;
    }
    super(message);
    Object.setPrototypeOf(this, CogniteError.prototype); // https://stackoverflow.com/questions/51229574/why-instanceof-returns-false-for-a-child-object-in-javascript
    this.errorMessage = errorMessage;
    this.status = status;
    this.requestId = requestId;
    this.missing = missing;
    this.duplicated = duplicated;
    this.extra = extra;
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    } else {
      this.stack = new Error(message).stack;
    }
  }

  public toJSON() {
    const jsonObject: { [key: string]: unknown } = {
      status: this.status,
      message: this.errorMessage,
    };
    if (this.requestId) {
      jsonObject.requestId = this.requestId;
    }
    if (this.missing) {
      jsonObject.missing = this.missing;
    }
    if (this.duplicated) {
      jsonObject.duplicated = this.duplicated;
    }
    if (this.extra) {
      jsonObject.extra = this.extra;
    }
    return jsonObject;
  }
}

/** @hidden */
export function handleErrorResponse(err: HttpError) {
  let code: number;
  let duplicated: object[] | undefined;
  let extra: unknown;
  let message: string;
  let missing: object[] | undefined;
  let requestId: string | undefined;
  try {
    code = err.status;
    const data = err.data;
    duplicated = data.error.duplicated;
    message = data.error.message;
    missing = data.error.missing;
    extra = data.error.extra;
    requestId = getHeaderField(err.headers || {}, X_REQUEST_ID);
  } catch (_) {
    throw err;
  }
  throw new CogniteError(message, code, requestId, {
    duplicated,
    extra,
    missing,
  });
}
