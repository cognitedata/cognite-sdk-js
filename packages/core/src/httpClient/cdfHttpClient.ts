// Copyright 2020 Cognite AS

import * as Url from 'url';
import {
  API_KEY_HEADER,
  AUTHORIZATION_HEADER,
  X_CDF_APP_HEADER,
  X_CDF_SDK_HEADER,
} from '../constants';
import { handleErrorResponse } from '../error';
import { bearerString, isJson } from '../utils';
import {
  HttpHeaders,
  HttpQueryParams,
  HttpRequest,
  HttpResponse,
} from './basicHttpClient';
import { HttpError } from './httpError';
import { RetryableHttpClient } from './retryableHttpClient';
import { RetryValidator } from './retryValidator';

export class CDFHttpClient extends RetryableHttpClient {
  private static serializeQueryParameters(
    params: HttpQueryParams = {}
  ): HttpQueryParams {
    return Object.keys(params).reduce(
      (serializedParams, key) => {
        const param = params[key];
        serializedParams[key] = isJson(param) ? JSON.stringify(param) : param;
        return serializedParams;
      },
      {} as HttpQueryParams
    );
  }

  private static isSameOrigin(baseUrl: string, url: string) {
    const { protocol: baseUrlProtocol, host: baseUrlHost } = Url.parse(baseUrl);
    const { protocol, host } = Url.parse(Url.resolve(baseUrl, url));
    const hasSameProtocol = baseUrlProtocol === protocol;
    const hasSameHost = baseUrlHost === host;
    return hasSameProtocol && hasSameHost;
  }

  private static filterHeaders(headers: HttpHeaders, names: string[]) {
    return names.reduce(
      (partiallyFilteredHeaders, headerName) =>
        CDFHttpClient.filterHeader(partiallyFilteredHeaders, headerName),
      headers
    );
  }

  private static filterHeader(headers: HttpHeaders, name: string) {
    const filteredHeaders: HttpHeaders = {
      ...headers,
    };
    delete filteredHeaders[name];
    return filteredHeaders;
  }

  private oneTimeHeaders: HttpHeaders = {};

  constructor(baseUrl: string, retryValidator: RetryValidator) {
    super(baseUrl, retryValidator);
  }

  public addOneTimeHeader(name: string, value: string) {
    this.oneTimeHeaders[name] = value;
    return this;
  }

  public setBearerToken = (token: string) => {
    this.setDefaultHeader(AUTHORIZATION_HEADER, bearerString(token));
  };

  public set401ResponseHandler(handler: Response401Handler) {
    this.response401Handler = handler;
  }

  protected async preRequest(request: HttpRequest): Promise<HttpRequest> {
    const headersWithDefaultHeaders = this.populateDefaultHeaders(
      request.headers
    );
    const headers = request.withCredentials
      ? headersWithDefaultHeaders
      : this.preventTokenLeakage(headersWithDefaultHeaders, request.path);
    const data = request.data;
    const params = CDFHttpClient.serializeQueryParameters(request.params);
    return {
      ...request,
      data,
      headers,
      params,
    };
  }

  protected async request<ResponseType>(request: HttpRequest) {
    request.headers = this.enrichWithOneTimeHeaders(request.headers);
    return super.request<ResponseType>(request);
  }

  protected async postRequest<T>(
    response: HttpResponse<T>,
    request: HttpRequest
  ): Promise<HttpResponse<T>> {
    try {
      return await super.postRequest(response, request);
    } catch (err) {
      if (
        err.status === 401 &&
        !this.isLoginOrLogoutApi(request.path) &&
        !this.isTokenInspect(request.path)
      ) {
        return new Promise((resolvePromise, rejectPromise) => {
          const retry = () => resolvePromise(this.request(request));
          const reject = () => rejectPromise(err);
          this.response401Handler(err, retry, reject);
        });
      }
      throw handleErrorResponse(err);
    }
  }

  private enrichWithOneTimeHeaders(headers?: HttpHeaders) {
    const disposableHeaders = this.oneTimeHeaders;
    this.oneTimeHeaders = {};
    return {
      ...headers,
      ...disposableHeaders,
    };
  }

  private response401Handler: Response401Handler = (_, __, reject) => reject();

  private preventTokenLeakage(headers: HttpHeaders, path: string) {
    if (CDFHttpClient.isSameOrigin(this.baseUrl, path)) {
      return headers;
    }
    return CDFHttpClient.filterHeaders(headers, [
      AUTHORIZATION_HEADER,
      API_KEY_HEADER,
      X_CDF_APP_HEADER,
      X_CDF_SDK_HEADER,
    ]);
  }

  private isLoginOrLogoutApi(url: string) {
    const lowerCaseUrl = url.toLowerCase();
    return (
      lowerCaseUrl.indexOf('/logout/url') !== -1 ||
      lowerCaseUrl.indexOf('/login/status') !== -1
    );
  }

  private isTokenInspect(url: string) {
    const lowerCaseUrl = url.toLowerCase();
    return lowerCaseUrl.indexOf('/token/inspect') !== -1;
  }
}

type Response401Handler = (
  err: HttpError,
  retry: () => void,
  reject: () => void
) => void;
