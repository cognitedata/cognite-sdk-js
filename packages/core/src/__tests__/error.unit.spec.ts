// Copyright 2020 Cognite AS
import { describe, expect, test } from 'vitest';

import { X_REQUEST_ID } from '../constants';
import { CogniteError, handleErrorResponse } from '../error';
import { HttpError } from '../httpClient/httpError';
import { createErrorResponse } from './testUtils';

const internalIdObject = { id: 4190022127342195 };
const externalIdObject = { externalId: 'abc' };
const event = {
  externalId: 'string',
  startTime: 0,
  endTime: 0,
  description: 'string',
  source: 'string',
};

describe('CogniteError', () => {
  test('without requestId', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const error = new CogniteError(errorMessage, status);
    expect(error.status).toBe(status);
    expect(error.requestId).toBeUndefined();
    expect(() => {
      throw error;
    }).toThrowErrorMatchingInlineSnapshot('[Error: Abc | code: 500]');
  });

  test('with requestId', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const requestId = 'def';
    const error = new CogniteError(errorMessage, status, requestId);
    expect(error.status).toBe(status);
    expect(error.requestId).toBe(requestId);
    expect(() => {
      throw error;
    }).toThrowErrorMatchingInlineSnapshot(
      '[Error: Abc | code: 500 | X-Request-ID: def]'
    );
  });

  test('extra field', () => {
    const errorMessage = 'Abc';
    const status = 500;
    const error = new CogniteError(errorMessage, status, undefined, {
      missing: [internalIdObject, externalIdObject],
      duplicated: [event],
      extra: 'Extra log information',
    });
    expect(() => {
      throw error;
    }).toThrowErrorMatchingSnapshot();
  });
});

describe('handleErrorResponse', () => {
  test('without requestId', () => {
    const httpError = new HttpError(500, createErrorResponse(500, 'abc'), {});
    expect(() => {
      handleErrorResponse(httpError);
    }).toThrowErrorMatchingInlineSnapshot('[Error: abc | code: 500]');
  });

  test('with requestId', () => {
    const httpError = new HttpError(500, createErrorResponse(500, 'abc'), {
      [X_REQUEST_ID]: 'def',
    });
    expect(() => {
      handleErrorResponse(httpError);
    }).toThrowErrorMatchingInlineSnapshot(
      '[Error: abc | code: 500 | X-Request-ID: def]'
    );
  });

  test('extra fields', () => {
    const status = 500;
    const message = 'abc';
    const xRequestId = 'def';
    const httpError = new HttpError(
      status,
      createErrorResponse(status, message, {
        missing: [internalIdObject, externalIdObject],
        duplicated: [event],
        extra: { extraErrorInformation: 'test' },
      }),
      { [X_REQUEST_ID]: xRequestId }
    );
    expect.assertions(6);
    expect(() => {
      handleErrorResponse(httpError);
    }).toThrowErrorMatchingSnapshot();

    try {
      handleErrorResponse(httpError);
    } catch (e) {
      if (!(e instanceof CogniteError)) {
        throw e;
      }
      expect(e.status).toBe(status);
      expect(e.requestId).toBe(xRequestId);
      expect(e.duplicated).toEqual([event]);
      expect((e.extra as Record<string, string>).extraErrorInformation).toBe(
        'test'
      );
      expect(e.missing).toEqual([internalIdObject, externalIdObject]);
    }
  });
});
