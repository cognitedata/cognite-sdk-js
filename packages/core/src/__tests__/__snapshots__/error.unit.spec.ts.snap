// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CogniteError extra field 1`] = `
"Abc | code: 500
{
  \\"missing\\": [
    {
      \\"id\\": 4190022127342195
    },
    {
      \\"externalId\\": \\"abc\\"
    }
  ],
  \\"duplicated\\": [
    {
      \\"externalId\\": \\"string\\",
      \\"startTime\\": 0,
      \\"endTime\\": 0,
      \\"description\\": \\"string\\",
      \\"source\\": \\"string\\"
    }
  ],
  \\"extra\\": \\"Extra log information\\"
}"
`;

exports[`handleErrorResponse extra fields 1`] = `
"abc | code: 500 | X-Request-ID: def
{
  \\"duplicated\\": [
    {
      \\"externalId\\": \\"string\\",
      \\"startTime\\": 0,
      \\"endTime\\": 0,
      \\"description\\": \\"string\\",
      \\"source\\": \\"string\\"
    }
  ],
  \\"extra\\": {
    \\"extraErrorInformation\\": \\"test\\"
  },
  \\"missing\\": [
    {
      \\"id\\": 4190022127342195
    },
    {
      \\"externalId\\": \\"abc\\"
    }
  ]
}"
`;
